README.md
# cargo-trend

[![Actions Status](https://github.com/dalance/cargo-trend/workflows/Regression/badge.svg)](https://github.com/dalance/cargo-trend/actions)
[![Crates.io](https://img.shields.io/crates/v/cargo-trend.svg)](https://crates.io/crates/cargo-trend)

cargo-trend is a cargo subcommand to generate trend graph of dependent crates.

## Installation

You can install by [cargo](https://crates.io/crates/cargo-trend).

```
cargo install cargo-trend
```

## Usage

Type `cargo trend` only. `trend.svg` show the trend graph of current project's dependencies.

```console
$ cargo trend
```

If crate names are given, `trend.svg` show the trend graph of the given crates.

```console
$ cargo trend failure error-chain quick-error snafu err-derive anyhow thiserror
```

`--output` option can specify the filename of the trend graph.
The extension can be `.svg`, `.png`, `.bmp`, `.jpeg` and `.jpg`.

```console
$ cargo trend --output trend.png
```

If `--relative` option is given, the y-axis of the output becomes fraction of crates.io.

If `--transitive` option is given, the way counting dependents changes from direct-dependencies to transitive-dependencies.

If `--top` option is given, the most trending crates are shown.

## Samples


### The most trending crates by 4 weeks

```console
$ cargo trend --top 5 --duration 4
```

![top5.svg](./samples/top5.svg)

```console
$ cargo trend --relative --top 5 --duration 4
```

![top5_relative.svg](./samples/top5_relative.svg)

### Error management library

```console
$ cargo trend failure error-chain quick-error snafu err-derive anyhow thiserror
```

![error.svg](./samples/error.svg)

```console
$ cargo trend --relative failure error-chain quick-error snafu err-derive anyhow thiserror
```

![error_relative.svg](./samples/error_relative.svg)

### Argument parser library

```console
$ cargo trend clap structopt docopt argparse getopts
```

![arg.svg](./samples/arg.svg)

```console
$ cargo trend --relative clap structopt docopt argparse getopts
```

![arg_relative.svg](./samples/arg_relative.svg)

### Terminal library

```console
$ cargo trend ansi_term termcolor term termion colored console
```

![term.svg](./samples/term.svg)

```console
$ cargo trend --relative ansi_term termcolor term termion colored console
```

![term_relative.svg](./samples/term_relative.svg)

### Async runtime library

```console
$ cargo trend tokio smol async-std
```

![async.svg](./samples/async.svg)

```console
$ cargo trend --relative tokio smol async-std
```

![async_relative.svg](./samples/async_relative.svg)

## Database

The database of dependent crates is `./db/db.gz`.
cargo-trend fetch it from github.com and generate graph.
The database is constructed from [crates.io-index](https://github.com/rust-lang/crates.io-index) and updated daily.


CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/dalance/cargo-trend/compare/v0.4.0...Unreleased) - ReleaseDate

## [v0.4.0](https://github.com/dalance/cargo-trend/compare/v0.4.0...v0.4.0) - 2023-05-18

## [v0.4.0](https://github.com/dalance/cargo-trend/compare/v0.3.0...v0.4.0) - 2023-05-18

* [Fixed] panic at invalid semver
* [Fixed] feature loop at DB update
* [Fixed] InvalidVersionRequirement error at DB update
* [Changed] update dependencies
* [Changed] DB version to v3

## [v0.3.0](https://github.com/dalance/cargo-trend/compare/v0.2.5...v0.3.0) - 2020-11-25

* [Fixed] wrong count of transitive
* [Fixed] feature trace at transitive count
* [Fixed] semver trace at transitive count

## [v0.2.5](https://github.com/dalance/cargo-trend/compare/v0.2.4...v0.2.5) - 2020-01-21

## [v0.2.4](https://github.com/dalance/cargo-trend/compare/v0.2.3...v0.2.4) - 2020-01-21

* [Fixed] build failure on Windows

## [v0.2.3](https://github.com/dalance/cargo-trend/compare/v0.2.2...v0.2.3) - 2020-01-16

## [v0.2.2](https://github.com/dalance/cargo-trend/compare/v0.2.1...v0.2.2) - 2020-01-16

* [Added] top option

## [v0.2.1](https://github.com/dalance/cargo-trend/compare/v0.2.0...v0.2.1) - 2020-01-15

## [v0.2.0](https://github.com/dalance/cargo-trend/compare/v0.1.1...v0.2.0) - 2020-01-15

* [Changed] DB version to v2
* [Added] transitive option
* [Added] relative option
* [Fixed] remove unwrap

## [v0.1.1](https://github.com/dalance/cargo-trend/compare/v0.1.0...v0.1.1) - 2019-10-04

* [Fixed] Fix subcommand parsing


src/db.rs
use anyhow::Error;
use chrono::serde::ts_seconds;
use chrono::{DateTime, TimeZone, Utc};
use crates_index::{Crate, Dependency, Index};
use dlhn::{Deserializer, Serializer};
use git2::{BranchType, Repository, ResetType};
use semver::{Version, VersionReq};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::{Read, Write};
use std::path::Path;
use tempfile::tempdir;

#[derive(Serialize, Deserialize, Debug)]
pub struct Db {
    pub update: DateTime<Utc>,
    pub map: HashMap<String, Vec<Entry>>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct DbHeader {
    pub update: DateTime<Utc>,
    pub hash: Vec<String>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct DbChunk {
    pub data: Vec<(String, Entry)>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Entry {
    #[serde(with = "ts_seconds")]
    pub time: DateTime<Utc>,
    pub direct_dependents: u64,
    pub transitive_dependents: u64,
    pub total_crates: u64,
}

impl Db {
    pub fn new() -> Db {
        Db {
            update: Utc.timestamp_opt(0, 0).unwrap(),
            map: HashMap::new(),
        }
    }

    pub fn load<T: AsRef<Path>>(dir: T) -> Result<Db, Error> {
        let path = dir.as_ref().join("db.json");
        let mut file = File::open(&path)?;
        let mut buf = Vec::new();
        file.read_to_end(&mut buf)?;
        let header: DbHeader = serde_json::from_str(&String::from_utf8(buf)?)?;

        let mut db = Db {
            update: header.update,
            map: HashMap::new(),
        };

        for i in 0..header.hash.len() {
            let path = dir.as_ref().join(format!("db{}", i));
            let mut file = File::open(path)?;
            let mut buf = Vec::new();
            file.read_to_end(&mut buf)?;
            let mut buf = buf.as_slice();
            let mut deserializer = Deserializer::new(&mut buf);
            let chunk = DbChunk::deserialize(&mut deserializer)?;

            for (name, entry) in chunk.data {
                db.map
                    .entry(name)
                    .and_modify(|e| e.push(entry.clone()))
                    .or_insert(vec![entry]);
            }
        }

        Ok(db)
    }

    pub fn save<T: AsRef<Path>>(&self, dir: T) -> Result<(), Error> {
        let mut map: Vec<_> = self.map.iter().collect();
        map.sort_by_key(|x| x.0);

        let mut data = Vec::new();
        for (k, v) in map {
            for e in v {
                data.push((k.to_owned(), e.to_owned()));
            }
        }
        data.sort_by_key(|x| x.1.time);

        let mut hashes = Vec::new();
        let mut i = 0;
        while data.len() > 1000000 {
            let rest = data.split_off(1000000);

            let path = dir.as_ref().join(format!("db{}", i));
            let hash = write_chunk(&path, data.clone())?;
            hashes.push(hash);

            data = rest;
            i += 1;
        }

        let path = dir.as_ref().join(format!("db{}", i));
        let hash = write_chunk(&path, data.clone())?;
        hashes.push(hash);

        let header = DbHeader {
            update: self.update.to_owned(),
            hash: hashes,
        };
        let encoded: Vec<u8> = serde_json::to_string(&header)?.into_bytes();
        let path = dir.as_ref().join("db.json");
        let mut file = File::create(path)?;
        file.write_all(&encoded)?;
        file.flush()?;

        Ok(())
    }

    pub fn update(&mut self, branch: Option<String>) -> Result<(), Error> {
        let url = "https://github.com/rust-lang/crates.io-index.git";
        let dir = tempdir()?;

        let repo = Repository::clone(url, &dir)?;
        repo.checkout_head(None)?;
        if let Some(branch) = branch {
            let master = repo.find_branch(&format!("origin/{}", branch), BranchType::Remote)?;
            let master = master.get().peel_to_commit()?;
            let master = repo.find_object(master.id(), None)?;
            repo.reset(&master, ResetType::Hard, None)?;
        }

        let mut revs = Vec::new();
        let mut revwalk = repo.revwalk()?;
        revwalk.push_head()?;
        let mut last = Utc.timestamp_opt(0, 0).unwrap();
        for oid in revwalk {
            let commit = repo.find_commit(oid?)?;
            let time = Utc.timestamp_opt(commit.time().seconds(), 0).unwrap();
            if last.date_naive() != time.date_naive() && time > self.update {
                revs.push((time, commit.id()));
                last = time;
            }
        }

        revs.reverse();

        let total = revs.len();
        for (i, (time, id)) in revs.iter().enumerate() {
            println!("Update DB: {} {} ( {} / {} )", time, id, i + 1, total);
            let obj = repo.find_object(id.clone(), None)?;
            repo.reset(&obj, ResetType::Hard, None)?;

            let index = Index::with_path(&dir.path(), crates_index::INDEX_GIT_URL)?;
            let mut crates = HashMap::new();
            for c in index.crates() {
                crates.insert(String::from(c.name()), c);
            }

            let mut deps = HashMap::new();
            let mut cache = HashMap::new();
            for (name, c) in &crates {
                let enabled_features = [String::from("default")];

                for dep in gather_dependencies(c, &VersionReq::STAR, &enabled_features) {
                    let name = dep.crate_name();
                    if let Some((cnt, _)) = deps.get_mut(name) {
                        *cnt += 1;
                    } else {
                        deps.insert(String::from(name), (1, 0));
                    }
                }

                let mut trace = HashSet::new();
                trace.insert(String::from(name));
                let (transitive, looped) = gather_transitive(
                    name,
                    &VersionReq::STAR,
                    &enabled_features,
                    trace,
                    &crates,
                    &mut cache,
                );

                // connect looped transitive
                for l in &looped {
                    let l_cached = cache.get(l).unwrap().clone();

                    for t in &transitive {
                        let t_cached = cache.get_mut(t).unwrap();
                        if t_cached.contains(l) {
                            for l in &l_cached {
                                t_cached.insert(l.clone());
                            }
                        }
                    }
                }

                for t in &transitive {
                    if let Some((_, cnt)) = deps.get_mut(t) {
                        *cnt += 1;
                    } else {
                        deps.insert(String::from(t), (0, 1));
                    }
                }
            }

            let total_crates = index.crates().count() as u64;
            for (name, (direct, transitive)) in &deps {
                let entry = Entry {
                    time: *time,
                    direct_dependents: *direct,
                    transitive_dependents: *transitive,
                    total_crates,
                };
                if let Some(entries) = self.map.get_mut(name) {
                    let last = &entries[entries.len() - 1];
                    if last.direct_dependents != *direct
                        || last.transitive_dependents != *transitive
                    {
                        (*entries).push(entry);
                    }
                } else {
                    self.map.insert(name.clone(), vec![entry]);
                }
            }

            self.update = *time;
        }

        Ok(())
    }

    pub fn fetch<T: AsRef<Path>>(dir: T) -> Result<(), Error> {
        let latest_header = reqwest::blocking::get(
            "https://raw.githubusercontent.com/dalance/cargo-trend/master/db_v3/db.json",
        )?
        .text()?;

        let header: DbHeader = serde_json::from_str(&latest_header)?;
        let path = dir.as_ref().join("db.json");
        let mut file = File::create(path)?;
        file.write_all(latest_header.as_bytes())?;
        file.flush()?;

        for (i, h) in header.hash.iter().enumerate() {
            let path = dir.as_ref().join(format!("db{}", i));
            let fetch = if path.exists() {
                let path = dir.as_ref().join(format!("db{}", i));
                let mut file = File::open(&path)?;
                let mut buf = Vec::new();
                file.read_to_end(&mut buf)?;
                let hash = format!("{:x}", Sha256::digest(&buf));
                &hash != h
            } else {
                true
            };

            if fetch {
                let mut res = reqwest::blocking::get(format!(
                    "https://github.com/dalance/cargo-trend/raw/master/db_v3/db{}",
                    i
                ))?;
                let mut buf = Vec::new();
                res.read_to_end(&mut buf)?;
                let path = dir.as_ref().join(format!("db{}", i));
                let mut file = File::create(&path)?;
                file.write_all(&buf)?;
                file.flush()?;
            }
        }

        Ok(())
    }
}

fn write_chunk(path: &Path, data: Vec<(String, Entry)>) -> Result<String, Error> {
    let chunk = DbChunk { data };

    let mut encoded = Vec::new();
    let mut serializer = Serializer::new(&mut encoded);
    chunk.serialize(&mut serializer)?;
    let mut file = File::create(&path)?;
    file.write_all(&encoded)?;
    file.flush()?;

    let mut file = File::open(&path)?;
    let mut buf = Vec::new();
    file.read_to_end(&mut buf)?;
    let hash = Sha256::digest(&buf);
    Ok(format!("{:x}", hash))
}

fn gather_transitive(
    name: &str,
    requirement: &VersionReq,
    enabled_features: &[String],
    trace: HashSet<String>,
    crates: &HashMap<String, Crate>,
    cache: &mut HashMap<String, HashSet<String>>,
) -> (HashSet<String>, HashSet<String>) {
    if let Some(cached) = cache.get(name) {
        (cached.clone(), HashSet::new())
    } else {
        let mut ret_looped = HashSet::new();
        let mut ret_transitive = HashSet::new();
        if let Some(c) = crates.get(name) {
            for dep in gather_dependencies(c, requirement, enabled_features) {
                let name = dep.crate_name();
                ret_transitive.insert(String::from(name));
                if !trace.contains(name) {
                    let mut trace = trace.clone();
                    trace.insert(String::from(name));

                    let mut dep_features: Vec<_> =
                        dep.features().iter().map(|x| x.clone()).collect();
                    if dep.has_default_features() {
                        dep_features.push(String::from("default"));
                    }

                    let requirement = match VersionReq::parse(dep.requirement()) {
                        Ok(x) => x,
                        Err(_) => continue,
                    };

                    let (transitive, looped) =
                        gather_transitive(name, &requirement, &dep_features, trace, crates, cache);
                    for l in looped {
                        ret_looped.insert(l.clone());
                    }
                    for t in transitive {
                        ret_transitive.insert(t.clone());
                    }
                } else {
                    ret_looped.insert(String::from(name));
                }
            }
        }
        cache.insert(String::from(name), ret_transitive.clone());
        (ret_transitive, ret_looped)
    }
}

fn gather_dependencies(
    krate: &Crate,
    requirement: &VersionReq,
    enabled_features: &[String],
) -> Vec<Dependency> {
    let krate = krate
        .versions()
        .iter()
        .filter(|x| {
            if let Ok(version) = Version::parse(x.version()) {
                requirement.matches(&version)
            } else {
                false
            }
        })
        .last();

    let mut ret = Vec::new();
    if let Some(krate) = krate {
        let enabled_deps = gather_enabled_dependencies(
            krate.features(),
            enabled_features,
            100,
            &mut HashSet::new(),
        );

        for dep in krate.dependencies() {
            if dep.is_optional() {
                if enabled_deps.iter().any(|x| x == dep.crate_name()) {
                    ret.push(dep.clone());
                }
            } else {
                ret.push(dep.clone());
            }
        }
    }

    ret
}

fn gather_enabled_dependencies(
    features: &HashMap<String, Vec<String>>,
    enabled_features: &[String],
    max_depth: usize,
    checked: &mut HashSet<String>,
) -> Vec<String> {
    let mut ret = Vec::new();
    for enabled in enabled_features {
        // break feature loop
        if checked.contains(enabled) {
            continue;
        }
        checked.insert(enabled.clone());

        if let Some(expanded) = features.get(enabled) {
            for e in expanded {
                let mut children = if max_depth == 0 {
                    Vec::new()
                } else {
                    gather_enabled_dependencies(features, &vec![e.clone()], max_depth - 1, checked)
                };
                if children.is_empty() {
                    ret.push(e.clone());
                } else {
                    ret.append(&mut children);
                }
            }
        }
    }
    ret
}


src/plotter.rs
use crate::db::Db;
use anyhow::Error;
use chrono::{NaiveDate, TimeZone, Utc};
use plotters::prelude::*;
use std::collections::BTreeMap;
use std::ffi::OsStr;
use std::path::Path;

pub struct Plotter {
    size: (u32, u32),
}

impl Plotter {
    pub fn new() -> Self {
        Plotter { size: (1200, 800) }
    }

    pub fn size(mut self, size: (u32, u32)) -> Self {
        self.size = size;
        self
    }

    pub fn plot<T: AsRef<Path>, U: AsRef<str>>(
        &self,
        path: T,
        targets: &[U],
        db: &Db,
        relative: bool,
        transitive: bool,
        start_date: Option<NaiveDate>,
    ) -> Result<(), Error> {
        let extension = path.as_ref().extension();
        match extension {
            Some(x) if x == OsStr::new("svg") => {
                let backend = SVGBackend::new(path.as_ref(), self.size);
                self.plot_with_backend(backend, targets, db, relative, transitive, start_date)
            }
            _ => {
                let backend = BitMapBackend::new(path.as_ref(), self.size);
                self.plot_with_backend(backend, targets, db, relative, transitive, start_date)
            }
        }
    }

    pub fn plot_with_backend<T, U>(
        &self,
        backend: T,
        targets: &[U],
        db: &Db,
        relative: bool,
        transitive: bool,
        start_date: Option<NaiveDate>,
    ) -> Result<(), Error>
    where
        T: DrawingBackend,
        T::ErrorType: 'static,
        U: AsRef<str>,
    {
        let mut x_min = Utc
            .timestamp_opt(std::i32::MAX as i64, 0)
            .unwrap()
            .date_naive();
        let mut x_max = Utc.timestamp_opt(0, 0).unwrap().date_naive();
        let mut y_min = std::f32::MAX;
        let mut y_max = std::f32::MIN;

        let mut plots = BTreeMap::new();
        for target in targets {
            let mut plot = Vec::new();
            if let Some(entries) = db.map.get(target.as_ref()) {
                for entry in entries {
                    let x_val = entry.time.date_naive();

                    if let Some(start) = start_date {
                        if start > x_val {
                            continue;
                        }
                    }

                    let dependents = if transitive {
                        entry.transitive_dependents
                    } else {
                        entry.direct_dependents
                    };
                    let y_val = if relative {
                        dependents as f32 / entry.total_crates as f32
                    } else {
                        dependents as f32
                    };
                    plot.push((x_val, y_val));

                    x_min = if x_min > x_val { x_val } else { x_min };
                    x_max = if x_max < x_val { x_val } else { x_max };
                    y_min = f32::min(y_min, y_val);
                    y_max = f32::max(y_max, y_val);
                }
            }
            plots.insert(String::from(target.as_ref()), plot);
        }

        y_min *= 0.9;
        y_max *= 1.1;

        let root = backend.into_drawing_area();
        let _ = root.fill(&WHITE);
        let root = root.margin(10, 10, 10, 10);
        let mut chart = ChartBuilder::on(&root)
            .x_label_area_size(50)
            .y_label_area_size(50)
            .build_cartesian_2d(x_min..x_max, y_min..y_max)?;

        let y_desc = if relative {
            "Fraction of dependent crates"
        } else {
            "Number of dependent crates"
        };

        chart
            .configure_mesh()
            .disable_x_mesh()
            .y_label_formatter(&|x| format!("{}", x))
            .y_desc(y_desc)
            .draw()?;

        let hue_step = 1.0 / plots.len() as f64;
        let mut hue = 0.0;
        for (target, plot) in &plots {
            let color = HSLColor(hue, 0.8, 0.5);
            hue += hue_step;

            let style = ShapeStyle {
                color: color.to_rgba(),
                filled: true,
                stroke_width: 2,
            };

            let anno = chart.draw_series(LineSeries::new(plot.clone(), style.clone()))?;
            anno.label(target).legend(move |(x, y)| {
                plotters::prelude::PathElement::new(vec![(x, y), (x + 20, y)], style.clone())
            });
        }

        chart
            .configure_series_labels()
            .position(SeriesLabelPosition::MiddleLeft)
            .background_style(&WHITE)
            .border_style(&BLACK)
            .draw()?;

        chart.plotting_area().present()?;
        Ok(())
    }
}


src/main.rs
mod db;
mod plotter;

use crate::db::Db;
use crate::plotter::Plotter;
use anyhow::{anyhow, Context, Error};
use cargo_metadata::MetadataCommand;
use chrono::{Duration, Utc};
use directories::ProjectDirs;
use std::fs;
use std::path::PathBuf;
use structopt::{clap, StructOpt};

// ---------------------------------------------------------------------------------------------------------------------
// Opt
// ---------------------------------------------------------------------------------------------------------------------

#[derive(Debug, StructOpt)]
#[structopt(bin_name = "cargo")]
#[structopt(setting(clap::AppSettings::DisableHelpSubcommand))]
pub enum CargoOpt {
    #[structopt(long_version(option_env!("LONG_VERSION").unwrap_or(env!("CARGO_PKG_VERSION"))))]
    #[structopt(setting(clap::AppSettings::ColoredHelp))]
    #[structopt(setting(clap::AppSettings::DeriveDisplayOrder))]
    Trend(Opt),
}

#[derive(Debug, StructOpt)]
pub struct Opt {
    /// Crates
    crates: Vec<String>,

    /// X size of output image
    #[structopt(value_name = "UINT", long = "xsize", default_value = "1200")]
    x_size: u32,

    /// Y size of output image
    #[structopt(value_name = "UINT", long = "ysize", default_value = "800")]
    y_size: u32,

    /// File path of output image
    #[structopt(
        value_name = "PATH",
        short = "o",
        long = "output",
        default_value = "trend.svg"
    )]
    output: PathBuf,

    /// File path of Cargo.toml
    #[structopt(value_name = "PATH", long = "manifest-path")]
    manifest_path: Option<PathBuf>,

    /// Update db
    #[structopt(value_name = "PATH", short = "u", long = "update")]
    update: Option<PathBuf>,

    /// Branch of crates.io-index
    #[structopt(value_name = "BRANCH", short = "b", long = "branch")]
    branch: Option<String>,

    /// Plot fraction of crates.io
    #[structopt(long = "relative")]
    relative: bool,

    /// Plot transitive dependents
    #[structopt(long = "transitive")]
    transitive: bool,

    /// The most trending crates
    #[structopt(
            value_name = "N",
            long = "top-trend",
            conflicts_with_all = &["top-dependent", "top_transitive"]
        )]
    top_trend: Option<usize>,

    /// The most dependent crates
    #[structopt(value_name = "N", long = "top-dependent", conflicts_with_all = &["top_trend", "top_transitive"])]
    top_dependent: Option<usize>,

    /// The most transitive dependent crates
    #[structopt(value_name = "N", long = "top-transitive", conflicts_with_all = &["top_trend", "top_dependent"])]
    top_transitive: Option<usize>,

    /// Duration by week
    #[structopt(long = "duration")]
    duration: Option<i64>,
}

// ---------------------------------------------------------------------------------------------------------------------
// Functions
// ---------------------------------------------------------------------------------------------------------------------

// ---------------------------------------------------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------------------------------------------------

fn main() {
    if let Err(e) = run() {
        let mut iter = e.chain();
        eprintln!("{}", iter.next().unwrap());
        for e in iter {
            eprintln!("  Caused by: {}", e);
        }

        std::process::exit(1);
    }
}

fn run() -> Result<(), Error> {
    let CargoOpt::Trend(opt) = CargoOpt::from_args();

    let mut transitive = opt.transitive;

    if let Some(path) = opt.update {
        let mut db = if path.exists() {
            Db::load(&path)?
        } else {
            Db::new()
        };
        db.update(opt.branch)?;
        db.save(&path)?;

        return Ok(());
    }

    let base_dir = ProjectDirs::from("org", "dalance", "cargo-trend")
        .ok_or_else(|| anyhow!("failed to find user directory"))?;
    let data_dir = base_dir.data_dir();
    fs::create_dir_all(data_dir).with_context(|| {
        format!(
            "failed to create data direcotry {}",
            data_dir.to_string_lossy()
        )
    })?;

    Db::fetch(data_dir)?;
    let db = Db::load(&data_dir)?;

    let start_date = if let Some(duration) = opt.duration {
        Some((Utc::now() - Duration::weeks(duration)).date_naive())
    } else {
        None
    };

    let targets = if let Some(top_trend) = opt.top_trend {
        let mut trend = Vec::new();
        for (name, entries) in &db.map {
            let mut entry_oldest = entries.first();
            for entry in entries {
                if let Some(start_date) = start_date {
                    if entry.time.date_naive() < start_date {
                        entry_oldest = Some(entry);
                    }
                }
            }
            let entry_newest = entries.last();

            if let Some(entry_oldest) = entry_oldest {
                if let Some(entry_newest) = entry_newest {
                    let (dep_oldest, dep_newest) = if transitive {
                        (
                            entry_oldest.transitive_dependents,
                            entry_newest.transitive_dependents,
                        )
                    } else {
                        (
                            entry_oldest.direct_dependents,
                            entry_newest.direct_dependents,
                        )
                    };

                    let (dep_oldest, dep_newest) = if opt.relative {
                        (
                            (dep_oldest as f64 / entry_oldest.total_crates as f64 * 10000.0) as i64,
                            (dep_newest as f64 / entry_newest.total_crates as f64 * 10000.0) as i64,
                        )
                    } else {
                        (dep_oldest as i64, dep_newest as i64)
                    };
                    trend.push((dep_newest - dep_oldest, name));
                }
            }
        }

        trend.sort_by_key(|x| x.0);

        let mut ret = Vec::new();
        for _ in 0..top_trend {
            if let Some(c) = trend.pop() {
                ret.push(c.1.clone());
            }
        }
        ret
    } else if let Some(top_dependent) = opt.top_dependent {
        transitive = false;

        let mut trend = Vec::new();
        for (name, entries) in &db.map {
            if let Some(entry) = entries.last() {
                trend.push((entry.direct_dependents, name));
            }
        }

        trend.sort_by_key(|x| x.0);

        let mut ret = Vec::new();
        for _ in 0..top_dependent {
            if let Some(c) = trend.pop() {
                ret.push(c.1.clone());
            }
        }
        ret
    } else if let Some(top_transitive) = opt.top_transitive {
        transitive = true;

        let mut trend = Vec::new();
        for (name, entries) in &db.map {
            if let Some(entry) = entries.last() {
                trend.push((entry.transitive_dependents, name));
            }
        }

        trend.sort_by_key(|x| x.0);

        let mut ret = Vec::new();
        for _ in 0..top_transitive {
            if let Some(c) = trend.pop() {
                ret.push(c.1.clone());
            }
        }
        ret
    } else if opt.crates.is_empty() {
        let mut cmd = MetadataCommand::new();
        if let Some(path) = opt.manifest_path {
            cmd.manifest_path(path);
        }
        let metadata = cmd.exec()?;

        let mut ret = Vec::new();
        for package in metadata.packages {
            if metadata
                .workspace_members
                .iter()
                .any(|x| x.repr == package.id.repr)
            {
                for dep in package.dependencies {
                    ret.push(dep.name);
                }
            }
        }
        ret
    } else {
        opt.crates
    };

    let plotter = Plotter::new().size((opt.x_size, opt.y_size));
    plotter.plot(
        opt.output,
        targets.as_slice(),
        &db,
        opt.relative,
        transitive,
        start_date,
    )?;

    Ok(())
}
